kimchi/src/circuits/polynomials/keccak/circuitgates.rs

```diff
@@ -0,0 +1,211 @@
+//~ <https://keccak.team/files/Keccak-reference-3.0.pdf>
+//~
+//~ |       | x = 3 | x = 4 | x = 0 | x = 1 | x = 2 |
+//~ | ----- | ----- | ----- | ----- | ----- | ----- |
+//~ | y = 2 |  155  |  231  |    3  |   10  |  171  |
+//~ | y = 1 |   55  |  276  |   36  |  300  |    6  |
+//~ | y = 0 |   28  |   91  |    0  |    1  |  190  |
+//~ | y = 4 |  120  |   78  |  210  |   66  |  253  |
+//~ | y = 3 |   21  |  136  |  105  |   45  |   15  |
+//~
+//~ ##### Design Approach:
+//~
+//~ The atomic operations are XOR, ROT, NOT, AND. In the sections below, we will describe
+//~ the gates for these operations. Below are some common approaches followed in their design.
+//~
+//~ To fit within 15 wires, we first decompose each word into its lower and upper 32-bit
+//~ components. A gate for an atomic operation works with those 32-bit components at a time.
+//~
+//~ Before we describe the specific gate design approaches, below are some constraints in the
+//~ Kimchi framework that dictated those approaches.
+//~ * only 4 lookups per row
+//~ * only first 7 columns are available to the permutation polynomial
+//~
+//~ ##### 32-bit decomposition gate
+//~
+//~ This is a basic operation that is typically done for 64-bit initial state and
+//~ intermediate values.
+//~
+//~ Let `inp` be a 64-bit word. The constraint is: `in` $= 2^32 \cdot$ `in_hi` $+$ `in_lo`.
+//~ It takes 3 cells for values `in`, `in_hi`, `in_lo`. We have not yet placed them w.r.t
+//~ other rows of the Keccak computation; the only requirement is that all these cells be
+//~ within the first 7 columns for permutation equation accessibility.
+//
+//~ ##### XOR gate
+//~
+//~ First we consider a XOR gate that checks that a 32-bit word `out` is the XOR of `in1` and `in2`.
+//~ This gate will use 2 rows, with a `Xor` row followed by a `Zero` row.
+//~
+//~ | Gates |          `Xor`   |          `Zero`  |
+//~ | ----- | ---------------- | ---------------- |
+//~ | Rows  |           0      |           1      |
+//~ | Cols  |                  |                  |
+//~ |     0 | copy     `in1`   | copy     `out`   |
+//~ |     1 |                  | copy     `in2`   |
+//~ |     2 |                  |                  |
+//~ |     3 | plookup0 `in2_0` | plookup4 `in2_4` |
+//~ |     4 | plookup1 `in2_1` | plookup5 `in2_5` |
+//~ |     5 | plookup2 `in2_2` | plookup6 `in2_6` |
+//~ |     6 | plookup3 `in2_3` | plookup8 `in2_7` |
+//~ |     7 | plookup0 `in1_0` | plookup4 `in1_4` |
+//~ |     8 | plookup1 `in1_1` | plookup5 `in1_5` |
+//~ |     9 | plookup2 `in1_2` | plookup6 `in1_6` |
+//~ |    10 | plookup3 `in1_3` | plookup7 `in1_7` |
+//~ |    11 | plookup0 `out_0` | plookup4 `out_4` |
+//~ |    12 | plookup1 `out_1` | plookup5 `out_5` |
+//~ |    13 | plookup2 `out_2` | plookup6 `out_6` |
+//~ |    14 | plookup3 `out_3` | plookup7 `out_7` |
+//~
+//~ Now we apply this gate twice to obtain a XOR gadget for 64-bit words by halving:
```

Comment:
Why are we leaving those empty spaces?  Might be good to say now.

Reply:
iirc it is pretty arbitrary